File: BinarySearchTree.java
public class BinarySearchTree {
    public static class Node {
        int data;
        Node left;
        Node right;

        public Node(int data) {
            this.data = data;
            this.left = null;
            this.right = null;
        }
    }

    public Node root;

    public BinarySearchTree() {
        root = null;
    }

    public void insert(int element) {
        Node newNode = new Node(element);

        //if tree is empty
        if (root == null) {
            root = newNode;
            return;
        } else {
            Node current = root, parent = null;

            while(true) {
                parent = current;

                if (element < current.data) {
                    current = current.left;
                    if (current == null) {
                        parent.left = newNode;
                        return;
                    }
                } else {
                    current = current.right;
                    if (current == null) {
                        parent.right = newNode;
                        return;
                    }
                }
            }
        }
    }

    public Node deleteNode(Node node, int value) {
        if (node == null)
            return null;
        else {
            if (value < node.data)
                node.left = deleteNode(node.left, value);
            else if (value > node.data)
                node.right = deleteNode(node.right, value);
            else {
                if (node.left == null && node.right == null)
                    node = null;
                else if (node.left == null)
                    node = node.right;
                else if (node.right == null)
                    node = node.left;
                else {
                    Node temp = minNode(node.right);
                    node.data = temp.data;
                    node.right = deleteNode(node.right, temp.data);
                }
            }
        }

        return node;
    }

    public Node minNode(Node root) {
        if (root.left!=null)
            return minNode(root.left);
        return root;
    }

    public void inorderTraversal(Node root) {
        if (root == null)
            return;
        inorderTraversal(root.left);
        System.out.print(root.data + " ");
        inorderTraversal(root.right);
    }

    public static void main(String[] args) {
        BinarySearchTree bst = new BinarySearchTree();
        
        bst.insert(7);
        bst.insert(14);
        bst.insert(6);
        bst.insert(7);
        bst.insert(2);
        bst.insert(18);
        bst.insert(14);

        bst.inorderTraversal(bst.root);
        System.out.println();

        bst.deleteNode(bst.root, 14);
        bst.inorderTraversal(bst.root);
        System.out.println();
    }
}

File: BinarySearchTree.py
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

class BinarySearchTree:
    def insert(self, node, val):
        # if tree empty
        if node is None:
            return Node(val)
        
        # traverse and insert
        if val < node.val:
            node.left = self.insert(node.left, val)
        else:
            node.right = self.insert(node.right, val)
        
        return node
    
    def delete(self, node, val):
        if node is None:
            return None
        else:
            if val < node.val:
                node.left = self.delete(node.left, val)
            elif val > node.val:
                node.right = self.delete(node.right, val)
            else:
                if node.left is None and node.right is None:
                    return None
                elif node.left is None:
                    node = node.right
                elif node.right is None:
                    node = node.left
                else:
                    temp = self.minNode(node.right)
                    node.val = temp.val
                    node.right = self.delete(node.right, temp.val)
        
        return node
    
    def minNode(self, root):
        if root.left is None:
            return root
        
        return self.minNode(root.left)
    
    def inorder(self, root):
        if root is None:
            return
        
        self.inorder(root.left)
        print(str(root.val) + "->", end=' ')
        self.inorder(root.right)
    
bst = BinarySearchTree()
root = None

# Insert nodes
values_to_insert = [50, 30, 20, 40, 70, 60, 80]
for val in values_to_insert:
    root = bst.insert(root, val)

print("In-order traversal before deletion:")
bst.inorder(root)
print("\n")

# Delete nodes
values_to_delete = [20, 30, 50]
for val in values_to_delete:
    root = bst.delete(root, val)

print("In-order traversal after deletion:")
bst.inorder(root)
print("\n")
